src/jit/importer.cpp

```diff
@@ -18269,4 +18271,325 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
 {
     return (tree->OperGet() == GT_INTRINSIC) && IsMathIntrinsic(tree->gtIntrinsic.gtIntrinsicId);
 }
-/*****************************************************************************/
+
+//------------------------------------------------------------------------
+// impDevirtualizeCall: Attempt to change a virtual vtable call into a
+//   normal call
+//
+// Arguments:
+//     call -- the call node to examine/modify
+//     obj  -- the value of 'this' for the call
+//     callInfo -- info about the call from the VM
+//     exactContextHnd -- [OUT] updated context handle iff call devirtualized
+//
+// Notes:
+//     Virtual calls in IL will always "invoke" the base class method.
+//
+//     This transformation looks for evidence that the type of 'this'
+//     in the call is a final class or would invoke a final method, and
+//     if that and other safety checks pan out, modifies the call and
+//     the call info to create a direct call.
+//
+//     This transformation is done in the importer and not in some
+//     subsequent optimization pass because we want it to be upstream
+//     of inline candidate identification.
+
+void Compiler::impDevirtualizeCall(GenTreeCall*            call,
+                                   GenTreePtr              obj,
+                                   CORINFO_CALL_INFO*      callInfo,
+                                   CORINFO_CONTEXT_HANDLE* exactContextHandle)
+{
+    // This should be a virtual vtable or virtual stub call.
+    assert(call->IsVirtual());
+
+    // Bail if not optimizing
+    if (opts.MinOpts())
+    {
+        return;
+    }
+
+    // Bail if debuggable codegen
+    if (opts.compDbgCode)
+    {
+        return;
+    }
+
+    // Bail if devirt is disabled.
+    if (JitConfig.JitEnableDevirtualization() == 0)
+    {
+        return;
+    }
+
+    // Do we know anything about the type of the 'this'?
```

Comment:
Yes, we will (always?) see an upstream Box and can work out the type from there. 

However I will probably leave this as future work since it would also be nice to optimize away the boxing at the same time.